From ed540bdbb6e5d4941e2c04530a44bbe00492e436 Mon Sep 17 00:00:00 2001
From: Leo Yan <leo.yan@intel.com>
Date: Mon, 19 Dec 2011 14:00:07 +0800
Subject: [PATCH] ltr502als.c: Proximity status check after mode switching in early suspend

BZ: 16573

If proximity status is changed during mode switch, there may be no
interrupt occurred, so we need to check the data register after mode
switching in the early suspend handler.

Change-Id: Iaf3e7d02a4709ee719dd14e3f05422b2a96c9c91
Signed-off-by: Leo Yan <leo.yan@intel.com>
---
 drivers/hwmon/ltr502als.c |    9 +++++++++
 1 files changed, 9 insertions(+), 0 deletions(-)

diff --git a/drivers/hwmon/ltr502als.c b/drivers/hwmon/ltr502als.c
index 7005ebd..b645bcc 100644
--- a/drivers/hwmon/ltr502als.c
+++ b/drivers/hwmon/ltr502als.c
@@ -600,6 +600,7 @@ static struct miscdevice ambient_dev = {
 #ifdef CONFIG_HAS_EARLYSUSPEND
 static void alsps_early_suspend(struct early_suspend *h)
 {
+	u8 data;
 	struct alsps_device *alsps = container_of(h, struct alsps_device, es);
 
 	dev_dbg(&alsps->client->dev, "enter %s\n", __func__);
@@ -607,6 +608,14 @@ static void alsps_early_suspend(struct early_suspend *h)
 	mutex_lock(&alsps_dev->lock);
 	/* Only proximity is kept actice over the suspend period */
 	ltr502_switch(alsps->alsps_switch & PROXIMITY_ENABLE);
+
+	/* If proximity status is changed during mode switch, there may be no
+	 * interrupt occurred, so we need to check the data register after mode
+	 * switching here */
+	/* Sleep 1 intergration cycle(100 ms) to ensure ADC sample */
+	msleep(100);
+	alsps_read(alsps, DATAREG, &data);
+	proximity_handle_irq(alsps, data);
 	mutex_unlock(&alsps_dev->lock);
 }
 
-- 
1.7.1

